lib/internal/repl/history.js

```diff
@@ -17,168 +25,405 @@ let debug = require('internal/util/debuglog').debuglog('repl', (fn) => {
+  canNavigateToNext() {
+    return this[kIndex] > -1 && this[kHistory].length > 0;
   }
 
-  let timer = null;
-  let writing = false;
-  let pending = false;
-  repl.pause();
-  // History files are conventionally not readable by others:
-  // https://github.com/nodejs/node/issues/3392
-  // https://github.com/nodejs/node/pull/3394
-  fs.open(historyPath, 'a+', 0o0600, oninit);
+  navigateToNext(substringSearch) {
+    if (!this.canNavigateToNext()) {
+      return null;
+    }
+    const search = substringSearch || '';
+    let index = this[kIndex] - 1;
+
+    while (
+      index >= 0 &&
+      (!StringPrototypeStartsWith(this[kHistory][index], search) ||
+        this[kContext].line === this[kHistory][index])
+    ) {
+      index--;
+    }
 
-  function oninit(err, hnd) {
-    if (err) {
-      // Cannot open history file.
-      // Don't crash, just don't persist history.
-      _writeToOutput(repl, '\nError: Could not open history file.\n' +
-        'REPL session history will not be persisted.\n');
-      debug(err.stack);
+    this[kIndex] = index;
 
-      repl._historyPrev = _replHistoryMessage;
-      repl.resume();
-      return ready(null, repl);
+    if (index === -1) {
+      return search;
     }
-    fs.close(hnd, onclose);
+
+    return ReplHistory[kNormalizeLineEndings](this[kHistory][index], '\r', '\n');
+  }
+
+  canNavigateToPrevious() {
+    return this[kHistory].length !== this[kIndex] && this[kHistory].length > 0;
   }
 
-  function onclose(err) {
-    if (err) {
-      return ready(err);
+  navigateToPrevious(substringSearch = '') {
+    if (!this.canNavigateToPrevious()) {
+      return null;
     }
-    fs.readFile(historyPath, 'utf8', onread);
+    const search = substringSearch || '';
+    let index = this[kIndex] + 1;
+
+    while (
+      index < this[kHistory].length &&
+      (!StringPrototypeStartsWith(this[kHistory][index], search) ||
+        this[kContext].line === this[kHistory][index])
+    ) {
+      index++;
+    }
+
+    this[kIndex] = index;
+
+    if (index === this[kHistory].length) {
+      return search;
+    }
+
+    return ReplHistory[kNormalizeLineEndings](this[kHistory][index], '\r', '\n');
   }
 
-  function onread(err, data) {
-    if (err) {
-      return ready(err);
+  get size() { return this[kSize]; }
+  get isFlushing() { return this[kIsFlushing]; }
+  get history() { return this[kHistory]; }
+  set history(value) { this[kHistory] = value; }
+  get index() { return this[kIndex]; }
+  set index(value) { this[kIndex] = value; }
+
+  // Start private methods
+
+  static [kGetHistoryPath](options) {
+    let historyPath = options.historyFile;
+    if (typeof historyPath === 'string') {
+      historyPath = StringPrototypeTrim(historyPath);
     }
+    return historyPath;
+  }
 
-    if (data) {
-      repl.history = RegExpPrototypeSymbolSplit(/\r?\n+/, data, repl.historySize);
-    } else {
-      repl.history = [];
+  static [kNormalizeLineEndings](line, from, to, reverse = true) {
```

Comment:
is the `reverse` parameter needed? it doesn't look to me like you ever pass `false` to it :thinking: 

(which makes me question the implementation below? is it actually correct? :thinking:  or maybe I'm missing something?)

Reply:
You are correct 😄 it looks like, while refactoring, I made it so that reversing it is always necessary, and I forgot to update the function 😄